rflx/pyrflx/typevalue.py

```diff
@@ -224,34 +237,221 @@ class OpaqueValue(TypeValue):
+        super().__init__(vtype)
+        self._element_type = vtype.element_type
+        self._is_message_array = isinstance(self._element_type, model.Message)
+
+    def assign(self, value: List[TypeValue], length: int = 0, check: bool = True) -> None:
+
+        # check if all elements are of the same class and messages are valid
+        for v in value:
+            if (
+                self._is_message_array
+                and isinstance(v, MessageValue)
+                and isinstance(self._element_type, model.Message)
+            ):
+                if not v.equal_model(self._element_type):
+                    raise ValueError("members of an array must not be of different classes")
+                if not v.valid_message:
+                    raise ValueError("cannot assign array of messages: messages must be valid")
+            else:
+                if not self._is_message_array and not v.equal_type(self._element_type):
+                    raise ValueError("members of an array must not be of different classes")
+
+        self._value = value
+
+    def assign_bitvalue(self, value: Bitstring, length: int = 0) -> None:
+
+        self._value = []
+
+        if self._is_message_array:
+
+            while len(str(value)) != 0:
+
+                nested_message = TypeValue.construct(self._element_type)
+                assert isinstance(nested_message, MessageValue)
+                try:
+                    nested_message.assign_bitvalue(value)
+                except Exception as e:
+                    raise ValueError(
+                        f"cannot parse nested messages in array of type "
+                        f"{self._element_type.full_name}: {e}"
+                    )
+                if nested_message.valid_message:
+                    self._value.append(nested_message)
+                else:
+                    raise ValueError(
+                        f"cannot append to array: message is invalid {nested_message.name}"
+                    )
+                value = value[len(str(nested_message.to_bitstring)) :]
+            return
+
+        if isinstance(self._element_type, model.Scalar):
+
+            value_str = str(value)
+            type_size = self._element_type.size
+            assert isinstance(type_size, Number)
+            type_size_int = type_size.value
 
-        self._value = bytestring
+            while len(value_str) != 0:
+                nested_value = TypeValue.construct(self._element_type)
+                nested_value.assign_bitvalue(Bitstring(value_str[:type_size_int]))
+                self._value.append(nested_value)
+                value_str = value_str[type_size_int:]
+            return
+
+        raise NotImplementedError(f"Arrays of {self._element_type} currently not supported")
 
     @property
     def length(self) -> int:
         self._raise_initialized()
-        return len(self._value) * 8
+        return len(str(self.to_bitstring))
```

Comment:
ScalarValues have a size property instead. The size property is computed by taking the size of the underling type and  returning the integer representation of it. As Opaque and Array (Types) do not have a previously known size i needed a way to determine the length of the OpaqueValue in order to know where the OpaqueValue field ends and the next field begins bitstring which is to be parsed.

Reply:
The calculation differs, but the meaning is the same. So they should be named consistently. It could probably make sense to add an abstract method to TypeValue.